client/app/bundles/course/reference-timelines/components/TimePopup/TimePopup.tsx

```diff
@@ -0,0 +1,164 @@
+import moment from 'moment';
+import {
+  ItemWithTimeData,
+  TimelineData,
+} from 'types/course/referenceTimelines';
+
+import useTranslation from 'lib/hooks/useTranslation';
+
+import { useSetLastSaved } from '../../contexts';
+import { createTime, deleteTime, updateTime } from '../../operations';
+import { useAppDispatch } from '../../store';
+import translations from '../../translations';
+import { DraftableTimeData } from '../../utils';
+import SeriouslyAnchoredPopup from '../SeriouslyAnchoredPopup';
+
+import TimePopupForm from './TimePopupForm';
+import TimePopupTopBar from './TimePopupTopBar';
+
+interface TimePopupProps {
+  for?: Partial<DraftableTimeData>;
+  assignedIn?: TimelineData;
+  assignedTo?: ItemWithTimeData;
+  anchorsOn?: HTMLElement;
+  default?: boolean;
+  gamified?: boolean;
+  newTime?: boolean;
+  onClose?: () => void;
+}
+
+const TimePopup = (props: TimePopupProps): JSX.Element => {
+  const {
+    anchorsOn: anchorElement,
+    for: time,
+    assignedIn: timeline,
+    assignedTo: item,
+  } = props;
+
+  const { t } = useTranslation();
+
+  const dispatch = useAppDispatch();
+
+  const { startLoading, abortLoading, setLastSavedToNow } = useSetLastSaved();
+
+  const handleCreateTime = (data: {
+    startAt: moment.Moment;
+    bonusEndAt?: moment.Moment;
+    endAt?: moment.Moment;
+  }): void => {
+    if (!timeline || !item) return;
+
+    startLoading();
+
+    dispatch(
+      createTime(timeline.id, item.id, {
+        startAt: data.startAt.toISOString(),
+        bonusEndAt: data.bonusEndAt?.toISOString(),
+        endAt: data.endAt?.toISOString(),
+      }),
+    )
+      .then(() => {
+        props.onClose?.();
+        setLastSavedToNow();
+      })
+      .catch((error) => {
+        abortLoading();
+        toast.error(error ?? t(translations.errorCreatingTime));
+      });
+  };
+
+  const handleUpdateTime = (data: {
+    startAt: moment.Moment;
+    bonusEndAt?: moment.Moment;
+    endAt?: moment.Moment;
+  }): void => {
+    if (!timeline || !item || !time?.id) return;
+
+    startLoading();
+
+    dispatch(
+      updateTime(timeline.id, item.id, time.id, {
+        startAt: data.startAt.toISOString(),
+        bonusEndAt: data.bonusEndAt?.toISOString() ?? null,
+        endAt: data.endAt?.toISOString() ?? null,
+      }),
+    )
+      .then(() => {
+        props.onClose?.();
+        setLastSavedToNow();
+      })
+      .catch((error) => {
+        abortLoading();
+        toast.error(error ?? t(translations.errorUpdatingTime));
+      });
+  };
+
+  const handleDeleteTime = (): void => {
+    if (!timeline || !item || !time?.id) return;
+
+    startLoading();
+
+    dispatch(deleteTime(timeline.id, item.id, time.id))
+      .then(() => {
+        props.onClose?.();
+        setLastSavedToNow();
+      })
+      .catch((error) => {
+        abortLoading();
+        toast.error(error ?? t(translations.errorDeletingTime));
+      });
+  };
+
+  return (
+    <SeriouslyAnchoredPopup
```

Comment:
That is expected behaviour of `Popover`. Even without `SeriouslyAnchoredPopup`, the popup will still appear at the leftmost part of the screen if the anchor element is either at the screen-left or off-screen to the left (and respectively on the right).

I thought about this before, but it would be great if `Popover` has a way to set bounds for its positioning. Otherwise, since `SeriouslyAnchoredPopup` already "hijacks" `left`/`top`, we could always put a `Math.min`/`Math.max` and set the bounds ourselves. The problem with this is, if we update our page layout (which we will, after SPA), we will have to change these values again; not an elegant solution.